Create StaticMetadataValue with substantial speed improvements [BUKKIT-1460]

crast opened the pull request:
This implementation yields from my test suites a ~20x speedup and ~1/2 the memory usage in metadata storage.

The original iteration of this (pull request #580) changed inheritance hierarchy, so I re-did this as creating a new base class. See https://bukkit.atlassian.net/browse/BUKKIT-1460 for more explanation.

This is two commits incoming because this also includes the changes from pull #771 which should be approved first before looking at this one.

To prove this is a worthy of creating a new class, I made a series of performance tests which I have in a separate branch: https://github.com/crast/Bukkit-Work/tree/static-metadata-perf-tests.

Running on my machine (Java SE 7, 64bit, OSX) I was able to get these results:

**Memory Test Results: (1 million objects)**

```
  FixedMetadataValue:
    phase1: 154.5 MB
    phase2: 198.3 MB

  StaticMetadataValue:
    phase1: 84.7 MB
    phase2: 84.8 MB
```

**Speed Test Results:**

```
Testing value FixedMetadataValue: 10000000 iterations
 => Elapsed 1670 ms
Testing value StaticMetadataValue: 10000000 iterations
 => Elapsed 82 ms
(~ 20.3x speedup )

Testing value FixedMetadataValue: 100000000 iterations
 => Elapsed 16783 ms
Testing value StaticMetadataValue: 100000000 iterations
 => Elapsed 841 ms
(~ 20.0x speedup)
```

As you can see, this uses significantly less memory (almost half), and provides a significant speedup (at the worst I saw a 3x speedup, and at some point it looks like HotSpot optimized it even more to a 20x speedup, while it seems HotSpot couldn't optimize FixedMetadataValue at all).

My recommendation would be to deprecate FixedMetadataValue and encourage people to use StaticMetadataValue (or at least provide a strong explanation in the documentation for FixedMetadataValue)

crast committed:
Refactor conversions from LazyMetadataValue into abstract base class

Implementing MetadataValue interface is significant work due to having
to provide a large amount of conversion stub method. This adds a new
optional abstract base class to aid in implementation.

Includes comprehensive unit tests including a sample adapter class, and
all existing metadata tests pass.

crast committed:
Created StaticMetadataValue to address BUKKIT-1460

Includes unit tests as well.

riking commented:
Note that there was a discussion at https://github.com/crast/Bukkit-Work/commit/4bf2954010d79d2848b20307bcbe1038fe08264b#commitcomment-2645909

riking commented on src/main/java/org/bukkit/metadata/StaticMetadataValue.java:
```diff
@@ -0,0 +1,38 @@
+        return this.internalValue;
+    }
+
+    /** Invalidating a StaticMetadataValue does nothing. */
```
Hm, should this throw a UnsupportedOperationException? That subclasses RuntimeException, by the way.

Would throwing that break some stuff? Actually, it's pretty hard to tell. How are we to know if some plugin goes around, invalidating every MetadataValue on some certain condition?

crast replied:
I'm fine with doing it either way. On one hand, because this is a new class, one could easily add new functionality that wasn't there in previous cases, but on the other hand, it could cause someone in the hypothetical "invalidate a bunch of values" case to start throwing exceptions where previously his code ran successfully.  Of course, that's a pretty easy to catch fix, though _technically_ it breaks API compatibility to start throwing an exception previously un-declared in the interface, even if it is a RuntimeException.

I sincerely doubt anyone goes around invalidating values not belonging to his own plugin though, because as it stands right now even LazyMetadataValue is not actually "lazy" (see my discussion on the topic here: http://goo.gl/u7m8B ) I have a pull request to address that by creating on-demand providers: https://github.com/Bukkit/Bukkit/pull/773

YLivay commented:
This would be so usefull. I would really love to see this in Bukkit :)

crast commented:
Addendum: I re-ran the perf tests on linux (x86_64, jdk7) and found a 15x speedup, and even more significant memory savings: https://gist.github.com/crast/5024885

spaceemotion commented:
Well thats a terrific amount of saved cycles and RAM usage - good job on this! I also would like to see this implemented.

crast committed:
Clean up whitespace.

amaranth commented:
Is there any reason you can't just change FixedMetadataValue to work like your proposed StaticMetadataValue? Sure, it'll still extend LazyMetadataValue but if you implement all of the methods and make them work like Static works you should get all of the performance gain. I would assume you'll still have a nice memory gain if not the same level you get now. I'd be willing to accept the increase in memory here to have a less confusing API.

Is this something you could try out and report the results?

crast commented:
@amaranth 

It will probably actually use slightly more memory, because it will have all the fields of LazyMetadataValue plus the static metadata field.  I'll do a test to find out, and I'll report on the results soon

crast commented:
@amaranth I made [a new branch](https://github.com/crast/Bukkit-Work/tree/alternate-fixedmetadata) with a new class, FixedMetadataValueEx, implemented on top of LazyMetadataValue just to see what would happen.

The result was speed improvements similar to StaticMetadataValue, but the memory improvements were minimal, as expected.

Memory usage (1 million objects):

```
FixedmetadataValue
 phase1: 155.7
 phase2: 192.4

StaticMetadataValue
 phase1: 84.8
 phase2: 84.7

FixedMetadataValueEx (Test overriding methods)
 phase1: 137.4
 phase2: 137.4
```

amaranth commented:
I think I'd prefer taking the memory hit for now to avoid confusion with the API. If we ever do another pile of breaking changes for API cleanup we could at that point simply remove the extends and be done with it.

crast commented:
@amaranth 
OK, Should I add commits to this pull request and let you guys squash them? Or should I make a new pull request?

natemort commented:
@crast We can squash the commits for you, go ahead and add them.

crast committed:
Re-factor StaticMetadataValue into FixedMetadataValue

It was requested that instead of using StaticMetadataValue as a new
class, that instead we'd fix FixedMetadataValue to get most of the
performance improvements, even if it came at a slight cost in memory
over the clean implementation.

amaranth commented:
Pulled, thanks.
